bazaar/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    //type Signature: Verify<Signer = Self::Public> + Member + Decode + Encode;
+}
+// Logger target
+const LOG: &str = "encointer";
+
+pub type ShopIdentifier = u64; //URL
+pub type ArticleIdentifier = u64; //URL
+/*
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, Default, RuntimeDebug)]
+pub struct ProofOfAttendance<Signature, AccountId> {
+    pub prover_public: AccountId,
+    pub ceremony_index: CeremonyIndexType,
+    pub currency_identifier: CurrencyIdentifier,
+    pub attendee_public: AccountId,
+    pub attendee_signature: Signature,
+}
+*/
+
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Bazaar {
+        // Maps the shop or article owner to the respective items 
+        pub ShopsOwned get(fn shops_owned): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) T::AccountId => Vec<ShopIdentifier>;
+        pub ArticlesOwned get(fn articles_owned): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) T::AccountId => Vec<ArticleIdentifier>; 
+        // Show item affiliation
+        pub ShopAffiliation get(fn shop_affiliation): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) ShopIdentifier => T::AccountId;
+        pub ArticleAffiliation get(fn article_affiliation): double_map hasher(blake2_128_concat) CurrencyIdentifier, hasher(blake2_128_concat) ArticleIdentifier => (T::AccountId, ShopIdentifier);
+        // The set of all shops and articles per currency (community)
+        pub ShopRegistry get(fn shop_registry): map hasher(blake2_128_concat) CurrencyIdentifier => Vec<ShopIdentifier>;
+        pub ArticleRegistry get(fn article_registry): map hasher(blake2_128_concat) CurrencyIdentifier => Vec<ArticleIdentifier>;
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where AccountId = <T as frame_system::Trait>::AccountId {
+        /// Event emitted when a shop is uploaded. [who, shop]
+        ShopCreated(AccountId, ShopIdentifier),
+        /// Event emitted when a shop is revoked by the owner. [who, shop]
+        ShopRemoved(AccountId, ShopIdentifier),
+    }
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// no such shop exisiting that could be deleted
+        NoSuchShop,
+        /// shop can not be created twice
+        ShopAlreadyCreated,
+        /// shop can not be removed by anyone else than its owner
+        OnlyOwnerCanRemoveShop,
+	}
+}
+
+// TODO: Check if URL valid?
+// TODO: Add Article Upload / Removal
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // initialisation
+      //  type Error = Error<T>;
+        fn deposit_event() = default;
+
+        /// Allow a user to create a shop
+        #[weight = 10_000]
+        pub fn upload_shop(origin, cid: CurrencyIdentifier, shop: ShopIdentifier) -> DispatchResult {
+            // Check that the extrinsic was signed and get the signer
+            let sender = ensure_signed(origin)?;
+
```

Comment:
you should verify that the cid supplied is a registered currency in `encointer_currencies` at this point